[ADD] payment_mollie: added mollie payment acquirer

pga-odoo opened the pull request:
Added mollie payment gateway. Mollie is a major payment provider in Europe and it is famous for providing multiple payments methods by a single acquirer. 

With mollie, you can use the following methods,
- Credit cards with inline component
- Apple Pay
- SOFORT banking
- SEPA Direct Debit
- SEPA Bank Transfer
- Paypal
- Klarna: Pay later
- Klarna: Slice it
- iDEAL
- Giropay
- EPS
- Bancontact
- KBC/CBC Payment Button
- Belfius Pay Button
- Eco-, gift- and meal vouchers
- Przelewy24
And many more

![image](https://user-images.githubusercontent.com/7629089/126712139-2801d488-4bfa-40e9-beec-af5a6279d37a.png)
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/74136).

pga-odoo commented:
Let me know if you have any queries :wink:

Yenthe666 commented:
Huh, this is interesting.. Last time I discussed & proposed it with Odoo (just over a year ago) they didn't want Mollie in and denied it. I'm curious what made them change their mind this time..

fmdl commented:
there are made odoo module him self : https://github.com/mollie/mollie-odoo

it is possible to backport in V14 ?

Yenthe666 commented:
@fmdl it is actually @pga-odoo that works on this module lately :) & I've already proposed it for V14 (even before V14 was released) and Odoo SA refused it so 🤷

AntoineVDV commented:
Hello, if this PR gets merged it will unfortunately not be backported to V14. This is mainly because the `payment` module was heavily refactored in 14.3 to make it easier to integrate new payment solutions. Integrating a payment provider in an earlier version is more complicated, is not always possible, and means supporting two very different versions of the same module.

AntoineVDV reviewed (changes requested):
Hello, @pga-odoo @kig-odoo first of all, kudos for writing clean and detailed code :clap:  It was really helpful for the review.

I had a deep look at this PR and came to the conclusion that a custom embedded payment method selection form is not the best way to integrate Mollie with Odoo. I believe a regular [Mollie Checkout](https://docs.mollie.com/payments/hosted-checkout) integration would offer almost the same payment options while requiring much fewer hacks and customizations. I understand that switching from one integration to another can be frustrating but it can also be exciting to achieve the same results with a fraction of the code.

From my point of view, the main problem with the proposed integration is that Mollie doesn't offer a pre-made widget for the selection of the payment method (PM) like Adyen or Stripe do. This makes it mandatory to go through a series of customizations that are generally best to avoid if we want to keep things simple and maintainable:
- synchronization of available PMs and their translations
- mixing of Mollie PMs and Mollie PM issuers in the `payment.icon` table
- mapping of Mollie PMs with Odoo's payment icons
- hack of the payment form to build a PM selection form and present PMs as acquirers
- implementation of issuers selection
This is *a lot* considering that pretty much all payment methods require the customer to be redirected to Mollie to finalize the payment.

I concede that switching to the Mollie Checkout integration also prevents from integrating more advanced features such as the configuration of vouchers or the integration with Sales, but we'd rather not have these in standard Odoo for now.

Note: The deadline for merging this PR has been pushed back one month to the 2nd of August.

pga-odoo commented:
@AntoineVDV Thank you for the feedback. We will start working on it and update the code based on **Mollie Checkout integration** and we will remove all extra things. 

Are we still want to use Order API and Payment API both or just Payment API? Order API is used in sales orders and enables some extra methods like Klarna Pay later 

Just putting this information here as an alternative, in case you like it, I can do something like this too. There is an alternative possible by adding a dropdown like the given image. And it depends on [Methods API](https://docs.mollie.com/reference/v2/methods-api/list-methods) to check available methods, so we can remove several things like the method table, issuer table, some JS code, PM selection form changes, and icons.  ![image](https://user-images.githubusercontent.com/7629089/127668139-a5a775d3-c39e-4342-af07-6dcbbbf8a6bb.png)

pga-odoo commented:
@AntoineVDV deadline push backed to 1 month so is the deadline is 2nd September or 2nd August?

AntoineVDV commented:
@pga-odoo Let's stick to the Payment API only for now. We see the Orders API as nice to have but a bit overkill. Moreover, integrating with Sales would require an additional module `sale_payment_mollie` as `payment_mollie` doesn't have any dependence on `sale`.

2nd of September*

pga-odoo commented:
@AntoineVDV thanks for the update.

I will let you know once it is completed according to your feedback.

pga-odoo commented:
@AntoineVDV I have updated the code as per your feedback. Now module is working based on **Mollie Checkout** 

I have also removed all extra code related to checkout customization and order API etc.  Compared to the previous implementation, code is reduced by 70% now :smile: 

Let me know if you want me to change anything else.

AntoineVDV commented on addons/payment_mollie/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+    'description': """Mollie Payment Acquirer""",
+
+    'author': 'Odoo S.A, Applix BV, Droggol Infotech Pvt. Ltd.',
+    'website': 'http://www.mollie.com',
```
```suggestion
    'website': 'https://www.mollie.com',
```

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+            # redirection, we include it in the redirect URL to be able to match the transaction.
+            'redirectUrl': f'{redirect_url}?ref={self.reference}',
+            'webhookUrl': f'{webhook_url}?ref={self.reference}',
+
```
```suggestion
```

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,166 @@
+    def _mollie_prepare_payment_payload(self):
+        """ This method prepare the payload based in api type.
+
+        Note: this method are splitted so we can write test cases
```
Alright, no need to this the form then if it has no input and if the action is not known in advance.

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+                "Mollie: " + _("Received data with invalid payment status: %s", payment_status)
+            )
+
+    def _mollie_prepare_payment_payload(self):
+        """ This method prepare the payload based in api type.
+
+        Note: this method are splitted so we can write test cases
+
+        :return: data of created record received from mollie api
+        :rtype: dict
+        """
```
```suggestion
    def _mollie_prepare_payment_request_payload(self):
        """ Create the payload for the payment request based on the transaction values.

        :return: The request payload
        :rtype: dict
        """
```
No need to specify that it is split for the tests. There is nothing wrong with extracting this method.

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+                "Mollie: " + _("Received data with invalid payment status: %s", payment_status)
+            )
+
+    def _mollie_prepare_payment_payload(self):
```
Please move this method after `_get_specific_rendering_values`.

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+        :return: data of created record received from mollie api
+        :rtype: dict
+        """
+        base_url = self.acquirer_id.get_base_url()
+        redirect_url = urls.url_join(base_url, MollieController._return_url)
+        user_lang = self.env.context.get('lang')
+        webhook_url = urls.url_join(base_url, MollieController._notify_url)
```
```suggestion
        user_lang = self.env.context.get('lang')
        base_url = self.acquirer_id.get_base_url()
        redirect_url = urls.url_join(base_url, MollieController._return_url)
        webhook_url = urls.url_join(base_url, MollieController._notify_url)
```

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+        user_lang = self.env.context.get('lang')
+        webhook_url = urls.url_join(base_url, MollieController._notify_url)
+
+        payment_data = {
```
```suggestion
        return {
```

AntoineVDV replied:
```suggestion
        return {
```

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+        if self.provider != 'mollie':
+            return res
+
+        api_url = False
```
```suggestion
```

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+            return res
+
+        api_url = False
+        mollie_payload = self._mollie_prepare_payment_payload()
```
```suggestion
        payload = self._mollie_prepare_payment_payload()
```

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,166 @@
+        # https://help.mollie.com/hc/en-us/articles/213470409
+        webhook_url = urls.url_join(base_url, MollieController._notify_url)
+        if "://localhost" not in webhook_url and "://192.168." not in webhook_url and "://127." not in webhook_url:
+            payment_data['webhookUrl'] = f'{webhook_url}?ref={self.reference}'
```
This is ok. Tests should always include a webhook anyway.

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,166 @@
+
+            # We are setting acquirer reference as we are receiving it before the redirection to mollie.
+            # This will help us verify the validity of the transaction later.
+            self.acquirer_reference = payment_data.get('id')
```
Alright, I didn't see that it was used for the request.

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+        if self.provider != 'mollie':
+            return
+
+        payment_data = self.acquirer_id._mollie_make_request(f'/payments/{self.acquirer_reference}', method="GET")
```
```suggestion
        payment_data = self.acquirer_id._mollie_make_request(
            f'/payments/{self.acquirer_reference}', method="GET"
        )
```

AntoineVDV commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,126 @@
+
+        api_url = False
+        mollie_payload = self._mollie_prepare_payment_payload()
+        payment_data = self.acquirer_id._mollie_make_request('/payments', data=mollie_payload)
```
```suggestion
        _logger.info("sending '/payments' request for link creation:\n%s", pprint.pformat(payload))
        payment_data = self.acquirer_id._mollie_make_request('/payments', data=mollie_payload)
```

Feyensv commented:
Note that the `ci/style` check should also be green on your PR before the merge (`ci/security` check is overridden manually after review from our internal security team).
It's red because of 5 unused imports remaining in your code ;) (Details [here](http://runbot.odoo.com/runbot/build/9673707))

AntoineVDV commented on addons/payment_mollie/controllers/main.py:
```diff
@@ -0,0 +1,51 @@
+import pprint
+
+from odoo import http
+from odoo.http import request, Response
+from odoo.exceptions import ValidationError
```
```suggestion
from odoo.exceptions import ValidationError
from odoo.http import request, Response
```

AntoineVDV commented on addons/payment_mollie/controllers/main.py:
```diff
@@ -0,0 +1,51 @@
+class MollieController(http.Controller):
+    _return_url = "/payment/mollie/return"
+    _notify_url = "/payment/mollie/notify"
+    @http.route(
```
```suggestion

    @http.route(
```

AntoineVDV commented on addons/payment_mollie/controllers/main.py:
```diff
@@ -0,0 +1,51 @@
+    )
+    def mollie_return(self, **data):
+        """ Process the data returned by Mollie after redirection.
+        The route is flagged with `save_session=False` to prevent Odoo from assigning a new session
```
```suggestion

        The route is flagged with `save_session=False` to prevent Odoo from assigning a new session
```

AntoineVDV commented on addons/payment_mollie/controllers/main.py:
```diff
@@ -0,0 +1,51 @@
+        request from the payment provider to Odoo. As the redirection to the '/payment/status' page
+        will satisfy any specification of the `SameSite` attribute, the session of the user will be
+        retrieved and with it the transaction which will be immediately post-processed.
+        :param dict data: The feedback data (only `id`) and the transaction reference (`ref`)
```
```suggestion

        :param dict data: The feedback data (only `id`) and the transaction reference (`ref`)
```

AntoineVDV commented on addons/payment_mollie/controllers/main.py:
```diff
@@ -0,0 +1,51 @@
+        :param dict data: The feedback data (only `id`) and the transaction reference (`ref`)
+                          embedded in the return URL
+        """
+
```
```suggestion
```

AntoineVDV commented on addons/payment_mollie/controllers/main.py:
```diff
@@ -0,0 +1,51 @@
+        request.env['payment.transaction'].sudo()._handle_feedback_data('mollie', data)
+        return request.redirect('/payment/status')
+
+    @http.route(_notify_url, type='http', auth='public', methods=['POST'], csrf=False, sitemap=False)
```
```suggestion
    @http.route(_notify_url, type='http', auth='public', methods=['POST'], csrf=False)
```

AntoineVDV commented on addons/payment_mollie/controllers/main.py:
```diff
@@ -0,0 +1,51 @@
+    @http.route(_notify_url, type='http', auth='public', methods=['POST'], csrf=False, sitemap=False)
+    def mollie_notify(self, **data):
+        """ Process the data sent by Mollie to the webhook.
+        :param dict data: The feedback data (only `id`) and the transaction reference (`ref`)
```
```suggestion

        :param dict data: The feedback data (only `id`) and the transaction reference (`ref`)
```

AntoineVDV commented on addons/payment_mollie/views/payment_views.xml:
```diff
@@ -0,0 +1,16 @@
+        <field name="arch" type="xml">
+            <xpath expr='//group[@name="acquirer"]' position='inside'>
+                <group attrs="{'invisible': [('provider', '!=', 'mollie')]}">
+                    <field name="mollie_api_key" attrs="{'required':[('provider', '=', 'mollie'), ('state', '!=', 'disabled')]}" password="True"/>
```
```suggestion
                    <field name="mollie_api_key" attrs="{'required': [('provider', '=', 'mollie'), ('state', '!=', 'disabled')]}" password="True"/>
```

AntoineVDV commented on addons/payment_mollie/tests/test_mollie.py:
```diff
@@ -0,0 +1,20 @@
+@tagged('post_install', '-at_install')
+class MollieTest(MollieCommon):
+
+    def test_payment_payload_with_method(self):
```
```suggestion
    def test_payment_request_payload_values(self):
```

AntoineVDV reviewed (changes requested):
Please have a look at the "CI/style" check at the bottom of the PR. There are a few imports to clean for the check to pass.

pga-odoo commented on addons/payment_mollie/const.py:
```diff
@@ -0,0 +1,21 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+# List of ISO 15897 locale supported by Mollie
```
it is  ISO 15897 locale. I have added mollies link which have locale details

pga-odoo commented on addons/payment_mollie/models/payment_acquirer.py:
```diff
@@ -0,0 +1,83 @@
+        string="API key",
+        help="The Test or Live API key depending on the configuration of the acquirer",
+        required_if_provider="mollie", groups="base.group_user"
+    )
```
I use `API key` because there is one field in the `payment_adyen` module which is also having the same label and it's giving this warning 

> Two fields (adyen_api_key, mollie_api_key) of payment.acquirer() have the same label: API Key

AntoineVDV replied:
Indeed, then please change the `string` in the field definition to `Mollie API Key` and to `API Key` in the view.

pga-odoo replied:
Done :+1:

AntoineVDV commented:
@pga-odoo Please reset my request for review when you're ready.

pga-odoo commented:
@AntoineVDV I have made changes according to your feedback.  I have added comments for some feedbacks. 

Waiting for runbot to pass all tests :crossed_fingers: 

Let me know if you need anything from my side :slightly_smiling_face:

Feyensv commented on addons/payment_mollie/models/payment_acquirer.py:
```diff
@@ -0,0 +1,82 @@
+        acquirers = super()._get_compatible_acquirers(*args, currency_id=currency_id, **kwargs)
+
+        currency = self.env['res.currency'].browse(currency_id).exists()
+        if currency:
+            acquirers = acquirers.filtered(
+                lambda a: a.provider != 'mollie' or currency.name in SUPPORTED_CURRENCIES
+            )
```
```suggestion
        if currency and currency.name not in SUPPORTED_CURRENCIES:
            acquirers = acquirers.filtered(lambda a: a.provider != 'mollie')
```
Move the generic condition outside the filtered lambda function, no need to evaluate it for each acquirer in the recordset ;).

Feyensv commented on addons/payment_mollie/models/payment_transaction.py:
```diff
@@ -0,0 +1,118 @@
+        Note: self.ensure_one() from `_get_processing_values`
+
+        :param dict processing_values: The generic and specific processing values of the transaction
+        :return: The dict of acquirer-specific processing values
```
```suggestion
        :return: The dict of acquirer-specific rendering values
```

pga-odoo commented:
@AntoineVDV @Feyensv tests are green now.
I have also updated everything according to your last feedback.

AntoineVDV reviewed (approved):
@pga-odoo Thank you for your quick work, it looks all good for me now :)

AntoineVDV commented:
@odoo/rd-security could you please review this one?

In particular, I'm not sure if the `except requests.exceptions.RequestException` (see CI logs) could this be an issue? We usually only catch specific subclasses and let other exceptions bubble up.

xmo-odoo commented:
> In particular, I'm not sure if the `except requests.exceptions.RequestException` (see CI logs) could be an issue? We usually only catch specific subclasses and let other exceptions bubble up.

It makes sense to intercept all failures as they're all, well, failures. Frankly at that point you might as well just `except Exception`.

The other endpoints are apparently more precise and do exclude some exception types but it seems mostly to provide different feedback for connection and http errors (why not) e.g. was it really a conscious decision to not handle `Timeout` errors and let those bubble up?

If you feel strongly about this it's easy enough to just catch the same two exceptions the other providers do, but I can't say I feel strongly about it.

pga-odoo commented:
@AntoineVDV Official python SDK from mollie is also using `except requests.exceptions.RequestException` [see here](https://github.com/mollie/mollie-api-python/blob/master/mollie/api/client.py#L215)

xmo-odoo commented:
@robodoo override=ci/security

usages of requests work fine, question is more in the details around it and that doesn't seem very security impacting.

AntoineVDV commented:
@robodoo r+

odony commented on addons/payment_mollie/models/payment_acquirer.py:
```diff
@@ -0,0 +1,80 @@
+        endpoint = f'/v2/{endpoint.strip("/")}'
+        url = urls.url_join('https://api.mollie.com/', endpoint)
+
+        odoo_version = service.common.exp_version()['server_version']
```
minor, but this is a complicated way of obtaining the `odoo.release.version` ;-)
```suggestion
        odoo_version = odoo.release.version
```
(after importing)

Feyensv commented:
@robodoo rebase-ff
